CODE_STYLE.md

```diff
@@ -0,0 +1,3293 @@
+<sup>[[link](#factory-methods)]</sup>
+
+  ```Ruby
+  class Person
+    def self.create(options_hash)
+      # body omitted
+    end
+  end
+  ```
+
+* <a name="duck-typing"></a>
+  Prefer [duck-typing](http://en.wikipedia.org/wiki/Duck_typing) over
+  inheritance.
+<sup>[[link](#duck-typing)]</sup>
+
+  ```Ruby
+  # bad
+  class Animal
+    # abstract method
+    def speak
+    end
+  end
+
+  # extend superclass
+  class Duck < Animal
+    def speak
+      puts 'Quack! Quack'
+    end
+  end
+
+  # extend superclass
+  class Dog < Animal
+    def speak
+      puts 'Bau! Bau!'
+    end
+  end
+
+  # good
+  class Duck
+    def speak
+      puts 'Quack! Quack'
+    end
+  end
+
+  class Dog
+    def speak
+      puts 'Bau! Bau!'
+    end
+  end
+  ```
+
+* <a name="no-class-vars"></a>
+  Avoid the usage of class (`@@`) variables due to their "nasty" behavior in
+  inheritance.
+<sup>[[link](#no-class-vars)]</sup>
+
+  ```Ruby
+  class Parent
+    @@class_var = 'parent'
+
+    def self.print_class_var
+      puts @@class_var
+    end
+  end
+
+  class Child < Parent
+    @@class_var = 'child'
+  end
+
+  Parent.print_class_var # => will print "child"
+  ```
+
+  As you can see all the classes in a class hierarchy actually share one
+  class variable. Class instance variables should usually be preferred
+  over class variables.
+
+* <a name="visibility"></a>
+  Assign proper visibility levels to methods (`private`, `protected`) in
+  accordance with their intended usage. Don't go off leaving everything `public`
+  (which is the default). After all we're coding in *Ruby* now, not in *Python*.
+<sup>[[link](#visibility)]</sup>
+
+* <a name="indent-public-private-protected"></a>
+  Indent the `public`, `protected`, and `private` methods as much as the method
+  definitions they apply to. Leave one blank line above the visibility modifier
+  and one blank line below in order to emphasize that it applies to all methods
+  below it.
+<sup>[[link](#indent-public-private-protected)]</sup>
+
+  ```Ruby
+  class SomeClass
+    def public_method
+      # ...
+    end
+
+    private
+
+    def private_method
+      # ...
+    end
+
+    def another_private_method
+      # ...
+    end
+  end
+  ```
+
+* <a name="def-self-class-methods"></a>
+  Use `def self.method` to define class methods. This makes the code
+  easier to refactor since the class name is not repeated.
+<sup>[[link](#def-self-class-methods)]</sup>
+
+  ```Ruby
+  class TestClass
+    # bad
+    def TestClass.some_method
+      # body omitted
+    end
+
+    # good
+    def self.some_other_method
+      # body omitted
+    end
+
+    # Also possible and convenient when you
+    # have to define many class methods.
+    class << self
+      def first_method
+        # body omitted
+      end
+
+      def second_method_etc
+        # body omitted
+      end
+    end
+  end
+  ```
+
+* <a name="alias-method-lexically"></a>
+  Prefer `alias` when aliasing methods in lexical class scope as the
```

Comment:
This point `use alias keyword` seems to be counter or at least in mild conflict with the next point about using `alias_method`